### DISCUSSION TITLE
What encoding is request.query_string?

### DISCUSSION THREAD

**Discussion by tucked** [pallets/werkzeug#3014] [Q&A]
Here it's encoded as `latin-1`:
https://github.com/pallets/werkzeug/blob/7868bef5d978093a8baa0784464ebe5d775ae92a/src/werkzeug/wrappers/request.py#L125

There are many instances where it's decoded as `utf-8`, though:
https://github.com/pallets/werkzeug/blob/7868bef5d978093a8baa0784464ebe5d775ae92a/src/werkzeug/sansio/request.py#L173

In my Flask-based app, decoding as `latin-1` works but is incorrect.

Why is it `bytes` anyway? I'm trying to get the original query string as a `str`.

**Comment by RonnyPfannschmidt** [comment]
- **Relevant standards**
  - [PEP 3333 – WSGI: Unicode Issues (native strings are latin-1 tunneled bytes)](https://peps.python.org/pep-3333/#unicode-issues)
  - [RFC 3986 §2.5 – URI Generic Syntax: Identifying Data (no encoding mandated)](https://www.rfc-editor.org/rfc/rfc3986#section-2.5)
  - [WHATWG URL Standard – application/x-www-form-urlencoded serializer (UTF-8 on write)](https://url.spec.whatwg.org/#concept-urlencoded-serializer)
  - [WHATWG URL Standard – application/x-www-form-urlencoded parser (UTF-8 on read)](https://url.spec.whatwg.org/#urlencoded-parsing)
- WSGI tunnels bytes through str as latin-1; werkzeug just undoes that.
- The URI spec mandates no encoding for the octets themselves.
- UTF-8 comes only from the URL Standard, which `request.args` follows.

**Reply by davidism** [reply]
Basically WSGI spec says that all values are strings, which was relevant in the Python 2-3 transition. But incoming HTTP data is bytes, so everything gets encoded as latin-1. That's a lossless encoding since it covers the full byte range, but can result in nonsense if it's not actually ASCII or latin-1 bytes. Therefore there's the whole "WSGI encoding dance" that we go through behind the scenes, which you point to in the code (`.encode("latin-1")`), where we have to encode the strings again in order to decode them with the proper codec.

For the query string, while it can be arbitrary bytes, that doesn't really happen in practice. We know it will be ASCII with `%` escapes for non-ASCII characters, and everyone uses UTF-8 now for non-ASCII.

The answer is, use:
- `request.args` if you want the parsed `?key=value`. This assumes ASCII plus percent-encoded UTF-8.
- `query_string` if you only care about the raw bytes.
- `query_string.decode(...)` if you only care about the raw _text_ and know its encoding.

As a side note, ASGI makes this a lot less complicated to implementers and people reading the code by specifying that everything remains raw bytes, and then it's up to the application to decode them appropriately. Rather than needing to re-encode them from string to bytes first like WSGI.
